aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java

```diff
@@ -393,6 +403,21 @@ public class S3FileIOProperties implements Serializable {
    */
   private static final String S3_FILE_IO_USER_AGENT = "s3fileio/" + EnvironmentContext.get();
 
+  /** Number of times to retry S3 operations. */
+  public static final String S3_RETRY_NUM_RETRIES = "s3.retry.num-retries";
+
+  public static final int S3_RETRY_NUM_RETRIES_DEFAULT = 32;
```

Comment:
@ookumuso This value seems really high to me? This can compound considering engine integrations may already have their own level of retries on top of this (e.g. Spark task retries). I would also hypothesize that the chance of success after some small  number of retries (e.g. I think the SDK default is 5?) is slim and that it's generally not worth trying beyond that. 

In the performed load tests did this 32 number markedly improved reliability generally compared to lower values? 

I think setting this default value comes down to what is the typical distribution of outcomes and I'd speculate that in the average case when there is 5xx errors this large number of retries won't help? In some extreme cases it seems possible, but I think for those cases, users can override the config defined here.

Reply:
@amogh-jahagirdar Main consideration here is to give S3 enough time to auto-scale particularly for workloads which observe 503s when they go beyond S3 advertised limits. What we observed in our load tests that high concurrent workloads exhaust the default retries(3) and fail immediately. How fast these retries exhausted will depend on the concurrency level. Retry count makes the main difference here since the overall application becomes more resistant to 503s which leads to success and better tps scaling for upcoming workloads with auto-scaling.

I would say in average case particularly workloads running below advertised limits, customers will not see a change in behavior since default retries wouldn't be exhausted for those anyway but as we see more adoption to Iceberg with bigger workloads running; exhausting retries on throttles becomes an issue. One downside would be that if there is a more systemic error in communications with S3, workloads will take longer time to fail, up to 10 minutes but I would argue that those will be more on the rare side compared to the value it brings for eventually succeeding workloads.

Reply:
@amogh-jahagirdar let me know whether you have any other questions

Reply:
>One downside would be that if there is a more systemic error in communications with S3, workloads will take longer time to fail, up to 10 minutes but I would argue that those will be more on the rare side compared to the value it brings for eventually succeeding workloads.

Thanks for your patience, yeah this is exactly the part I'm mulling over. I can see a rationale for retrying specifically throttling this many times for larger workloads but if I'm reading this PR right, this number of retries is for all retryable exceptions, so I'm double checking if this large level of retries is too broad for all the possible retryable exceptions.

Reply:
I would agree with the concern here.  If we can narrow this retry behavior to just the 503 error code, I would be more amenable (though 32 is still very high), but as a default retry this causes really bad behaviors for other errors.  The other issue is that there are retries in the surrounding execution path (task retries and stage retries in spark for example).  The compounding effect of high retry values has a multiplicative effect for the overall job.

I would also say that trying to hide this is problematic for those diagnosing slowness with their workloads, so we should log messages when slowdowns are occurring.

Reply:
@amogh-jahagirdar @danielcweeks See the response from Drew [here](https://github.com/apache/iceberg/pull/11052#discussion_r1777451480).

> If we can narrow this retry behavior to just the 503 error code

Token bucket behavior should exactly achieve that since it captures all non-throttling exceptions and disables retries.